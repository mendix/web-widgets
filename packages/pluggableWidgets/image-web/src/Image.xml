<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.web.image.Image" needsEntityContext="false" offlineCapable="true" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd" pluginWidget="true">
    <name>Image</name>
    <description>Display an image and enlarge it on click.</description>
    <studioProCategory>Images, videos &amp; files</studioProCategory>
    <studioCategory>Images, Videos &amp; Files</studioCategory>
    <helpUrl>https://docs.mendix.com/appstore/widgets/image</helpUrl>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="Data source">
                <property key="datasource" type="enumeration" defaultValue="image">
                    <caption>Image type</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="image">Image</enumerationValue>
                        <enumerationValue key="imageUrl">Image URL</enumerationValue>
                        <enumerationValue key="icon">Icon</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="imageObject" type="image" required="false">
                    <caption>Image source</caption>
                    <description />
                </property>
                <property key="defaultImageDynamic" type="image" required="false">
                    <caption>Default image</caption>
                    <description>This is the image that is displayed if no image is uploaded.</description>
                </property>
                <property key="imageUrl" type="textTemplate" required="false">
                    <caption>Image URL</caption>
                    <description>The link of the external image.</description>
                </property>
                <property key="imageIcon" type="icon" required="false">
                    <caption>Icon</caption>
                    <description>The icon image.</description>
                </property>
                <property key="isBackgroundImage" type="boolean" defaultValue="false">
                    <caption>Background image</caption>
                    <description>Whether the image is rendered as a background. More content can be put inside, while design properties will have no effect.</description>
                </property>
                <property key="children" type="widgets" required="false">
                    <caption>Place content here</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Events">
                <property key="onClickType" type="enumeration" defaultValue="action">
                    <caption>On click type</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="action">Action</enumerationValue>
                        <enumerationValue key="enlarge">Enlarge</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="onClick" type="action" required="false">
                    <caption>On click</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Accessibility">
                <property key="alternativeText" type="textTemplate" required="false">
                    <caption>Alternative text</caption>
                    <description>Alternative text of the image for accessibility purposes.</description>
                </property>
            </propertyGroup>
            <propertyGroup caption="Conditional Visibility">
                <systemProperty key="Name" />
                <systemProperty key="Visibility" />
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Dimensions">
            <propertyGroup caption="Dimensions">
                <property key="widthUnit" type="enumeration" defaultValue="auto">
                    <caption>Width unit</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="auto">Auto</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="width" type="integer" defaultValue="100">
                    <caption>Width</caption>
                    <description />
                </property>
                <property key="heightUnit" type="enumeration" defaultValue="auto">
                    <caption>Height unit</caption>
                    <description>Auto will keep the aspect ratio of the image.</description>
                    <enumerationValues>
                        <enumerationValue key="auto">Auto</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                        <enumerationValue key="viewport">Viewport</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="height" type="integer" defaultValue="100">
                    <caption>Height</caption>
                    <description />
                </property>
                <property key="minHeightUnit" type="enumeration" defaultValue="none">
                    <caption>Minimum Height unit</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                        <enumerationValue key="viewport">Viewport</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="minHeight" type="integer" required="true" defaultValue="0">
                    <caption>Minimum height</caption>
                    <description />
                </property>
                <property key="maxHeightUnit" type="enumeration" defaultValue="none">
                    <caption>Maximum Height unit</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="none">None</enumerationValue>
                        <enumerationValue key="pixels">Pixels</enumerationValue>
                        <enumerationValue key="percentage">Percentage</enumerationValue>
                        <enumerationValue key="viewport">Viewport</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="maxHeight" type="integer" required="true" defaultValue="0">
                    <caption>Maximum height</caption>
                    <description />
                </property>
                <property key="iconSize" type="integer" defaultValue="14">
                    <caption>Icon size</caption>
                    <description>The size of the icon in pixels.</description>
                </property>
                <property key="displayAs" type="enumeration" defaultValue="fullImage">
                    <caption>Show</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="fullImage">Full image</enumerationValue>
                        <enumerationValue key="thumbnail">Thumbnail</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="responsive" type="boolean" defaultValue="true">
                    <caption>Responsive</caption>
                    <description>Image will never get larger than its original size. It can become smaller.</description>
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
