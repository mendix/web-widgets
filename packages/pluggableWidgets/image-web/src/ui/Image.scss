.mx-image-viewer {
    width: 100%;
}

.mx-image-viewer-responsive img {
    max-width: 100%;
    max-height: 100%;
    display: block;
}

.mx-image-viewer-lightbox-backdrop {
    // Should be higher than the z-index of `.layout-atlas .region-sidebar` so it sits on top of the page.
    z-index: 110;
    background-color: rgba(0, 0, 0, 0.85);
    position: fixed;
    inset: 0px;

    .close-button {
        position: absolute;
        top: 30px;
        right: 30px;
        color: white;
        display: flex;
        background: none;
        border: none;
        padding: 8px;
        cursor: pointer;

        .removeIcon {
            height: 16px;
            width: 16px;
        }
    }
}

.mx-image-viewer-lightbox {
    // Should be higher than the z-index of `.mx-image-viewer-lightbox-backdrop`.
    z-index: 210;
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
}

.mx-image-viewer.mx-image-background {
    background-repeat: no-repeat;
    background-position: center;
}
