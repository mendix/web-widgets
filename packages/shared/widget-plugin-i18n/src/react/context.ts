import { createContext } from "react";
import { PluginI18n } from "../types";

const contextPath = "com.mendix.widget.plugin.i18n.context" as const;

type I18nContext = React.Context<PluginI18n | null>;

const createI18nContext = (): I18nContext => createContext<PluginI18n | null>(null);

declare global {
    interface Window {
        [contextPath]?: ReturnType<typeof createI18nContext>;
    }
}

export function getGlobalI18nPluginContext(): I18nContext {
    return window[contextPath] ?? createI18nContext();
}
