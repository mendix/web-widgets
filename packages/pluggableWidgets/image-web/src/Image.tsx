import { ValueStatus } from "mendix";
import { createElement, FunctionComponent, useCallback } from "react";
import { ImageContainerProps } from "../typings/ImageProps";
import { Image as ImageComponent, ImageType } from "./components/Image/Image";
import { constructStyleObject } from "./utils/helpers";

function getImageProps({
    datasource,
    imageIcon,
    imageObject,
    imageUrl,
    defaultImageDynamic
}: ImageContainerProps): ImageType {
    const fallback: ImageType = {
        type: "image",
        image: undefined
    };
    switch (datasource) {
        case "image": {
            if (imageObject?.status === ValueStatus.Available) {
                return {
                    type: "image",
                    image: imageObject.value.uri
                };
            } else if (
                imageObject?.status === ValueStatus.Unavailable &&
                defaultImageDynamic?.status === ValueStatus.Available
            ) {
                return {
                    type: "image",
                    image: defaultImageDynamic.value.uri
                };
            }
            return {
                type: "image",
                image: undefined
            };
        }
        case "imageUrl":
            return {
                type: "image",
                image: imageUrl?.status === ValueStatus.Available ? imageUrl.value : undefined
            };
        case "icon": {
            if (imageIcon?.status === ValueStatus.Available && imageIcon.value) {
                const icon = imageIcon.value;
                return {
                    type: icon.type,
                    image: icon.type === "image" ? icon.iconUrl : icon.iconClass
                };
            }
            return fallback;
        }
        default:
            return fallback;
    }
}

export const Image: FunctionComponent<ImageContainerProps> = props => {
    const onClick = useCallback(() => props.onClick?.execute(), [props.onClick]);
    const { type, image } = getImageProps(props);

    const altText = props.alternativeText?.status === ValueStatus.Available ? props.alternativeText.value : undefined;
    const styleObject = constructStyleObject(props);

    const imageStyle = { ...props.style, ...styleObject };
    console.warn("Image styling debug:", {
        name: props.name,
        heightUnit: props.heightUnit,
        minHeightUnit: props.minHeightUnit,
        minHeight: props.minHeight,
        styleObject,
        finalImageStyle: imageStyle
    });

    return (
        <ImageComponent
            class={props.class}
            style={imageStyle}
            widthUnit={props.widthUnit}
            width={props.width}
            heightUnit={props.heightUnit}
            height={props.height}
            tabIndex={props.tabIndex}
            iconSize={props.iconSize}
            responsive={props.responsive}
            onClickType={props.onClickType}
            onClick={props.onClick ? onClick : undefined}
            type={type}
            image={image}
            altText={altText}
            displayAs={props.displayAs}
            renderAsBackground={props.datasource !== "icon" && props.isBackgroundImage}
            backgroundImageContent={props.children}
        />
    );
};
