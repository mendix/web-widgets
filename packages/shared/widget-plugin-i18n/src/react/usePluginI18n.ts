import { useState } from "react";
import { I18nController } from "../controllers/I18nController";
import { PluginI18n } from "../types";

export function usePluginI18n(languageCode: string): PluginI18n | null {
    const [plugin, setPlugin] = useState<PluginI18n | null>(() => {
        new I18nController({
            languageCode,
            onPluginChange: plugin => setPlugin(plugin)
        });

        return null;
    });

    return plugin;
}
