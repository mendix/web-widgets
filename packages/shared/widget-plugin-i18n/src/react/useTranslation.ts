import { useContext } from "react";
import { PluginI18n } from "../types";
import { getGlobalI18nPluginContext } from "./context";

export function useTranslation(): PluginI18n | null {
    return useContext(getGlobalI18nPluginContext());
}
