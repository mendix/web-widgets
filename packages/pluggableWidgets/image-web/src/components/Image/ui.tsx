import { CSSProperties, HTMLAttributes, ReactElement, ReactEventHandler, ReactNode } from "react";
import classNames from "classnames";
import { HeightUnitEnum, WidthUnitEnum } from "../../../typings/ImageProps";
import { LightboxProps } from "../Lightbox";

import "../../ui/Image.scss";

function getStyle(value: string | number, type: WidthUnitEnum | HeightUnitEnum): number | string {
    // when type is auto default browser styles applies
    if (type === "pixels") {
        return value;
    } else if (type === "percentage") {
        return value + "%";
    }

    return "";
}
export interface ImageWrapperProps {
    className?: string;
    responsive: boolean;
    hasImage: boolean;
    children:
        | ReactElement<ImageContentIcon | ImageContentImage>
        | [ReactElement<ImageContentIcon | ImageContentImage>, ReactElement<LightboxProps> | false];
}

export interface ImageContentProps {
    style?: CSSProperties;
    tabIndex?: number;
    onClick?: ReactEventHandler<HTMLElement>;
    altText?: string;
}

function Wrapper(props: ImageWrapperProps): ReactElement {
    return (
        <div
            className={classNames(
                "mx-image-viewer",
                { "mx-image-viewer-responsive": props.responsive },
                props.className,
                { hidden: !props.hasImage }
            )}
        >
            {props.children}
        </div>
    );
}

export interface ImageContentIcon extends ImageContentProps {
    icon: string | undefined;
    size: number;
    isGlyph?: boolean;
}

function ContentIcon(props: ImageContentIcon): ReactElement {
    const accessibilityProps = props.altText
        ? {
              "aria-label": props.altText,
              role: "img"
          }
        : {
              role: "img"
          };

    const onClickProps = getImageContentOnClickProps(props.onClick, props.tabIndex);

    return (
        <span
            className={classNames(props.icon, { glyphicon: props.isGlyph })}
            style={{ ...props.style, fontSize: `${props.size}px` }}
            {...accessibilityProps}
            {...onClickProps}
        />
    );
}

export interface ImageContentImage extends ImageContentProps {
    image: string | undefined;
    height: number;
    heightUnit: HeightUnitEnum;
    width: number;
    widthUnit: WidthUnitEnum;
    className?: string;
}

function ContentImage(props: ImageContentImage): ReactElement {
    const onClickProps = getImageContentOnClickProps(props.onClick, props.tabIndex);
    return (
        <img
            className={props.className}
            src={props.image}
            style={{
                ...props.style,
                height: getStyle(props.height, props.heightUnit),
                width: getStyle(props.width, props.widthUnit)
            }}
            alt={props.altText}
            {...onClickProps}
        />
    );
}

function getImageContentOnClickProps(
    onClick: ImageContentProps["onClick"],
    tabIndex?: number
): HTMLAttributes<HTMLElement> {
    if (!onClick) {
        return {
            role: "img"
        };
    }
    return {
        onClick,
        role: "button",
        tabIndex: tabIndex ?? 0,
        onKeyDown: event => {
            if (event.key === "Enter" || event.key === " ") {
                onClick(event);
            }
        }
    };
}

interface BackgroundImageProps extends ImageContentImage {
    children: ReactNode;
}

function BackgroundImage(props: BackgroundImageProps): ReactElement {
    return (
        <div
            className={classNames("mx-image-viewer", "mx-image-background", props.className)}
            onClick={props.onClick}
            style={{
                ...props.style,
                height: getStyle(props.height, props.heightUnit),
                width: getStyle(props.width, props.widthUnit),
                backgroundImage: `url('${props.image}')`
            }}
        >
            {props.children}
        </div>
    );
}

export const ImageUi = {
    Wrapper,
    BackgroundImage,
    ContentIcon,
    ContentImage
};
