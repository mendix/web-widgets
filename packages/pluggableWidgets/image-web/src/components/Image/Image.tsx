import { cloneElement, CSSProperties, FunctionComponent, ReactNode, useCallback } from "react";
import { DisplayAsEnum, HeightUnitEnum, OnClickTypeEnum, WidthUnitEnum } from "../../../typings/ImageProps";
import { useLightboxState } from "../../utils/lightboxState";
import { ImageContentProps, ImageUi } from "./ui";
import { Lightbox, LightboxProps } from "../Lightbox";

import "../../ui/Image.scss";
import classNames from "classnames";

export type ImageType = {
    type: "image" | "icon" | "glyph";
    image: string | undefined;
};

export interface ImageProps extends ImageType {
    class: string;
    style?: CSSProperties;
    widthUnit: WidthUnitEnum;
    width: number;
    heightUnit: HeightUnitEnum;
    height: number;
    tabIndex?: number;
    iconSize: number;
    responsive: boolean;
    onClickType: OnClickTypeEnum;
    onClick?: () => void;
    altText?: string;
    displayAs: DisplayAsEnum;
    previewMode?: boolean;
    renderAsBackground: boolean;
    backgroundImageContent?: ReactNode;
}

function processImageLink(imageLink: string | undefined, displayAs: DisplayAsEnum): string | undefined {
    if (!imageLink || displayAs === "fullImage") {
        return imageLink;
    }
    const url = new URL(imageLink);
    url.searchParams.append("thumb", "true");
    return url.href;
}

export const Image: FunctionComponent<ImageProps> = ({
    class: className,
    style,
    widthUnit,
    width,
    heightUnit,
    height,
    tabIndex,
    iconSize,
    responsive,
    onClickType,
    onClick,
    type,
    image,
    altText,
    displayAs,
    previewMode,
    renderAsBackground,
    backgroundImageContent
}) => {
    const { lightboxIsOpen, openLightbox, closeLightbox } = useLightboxState();

    const onCloseLightbox = useCallback<LightboxProps["onClose"]>(
        event => {
            event?.stopPropagation();
            closeLightbox();
        },
        [closeLightbox]
    );

    const onImageClick = useCallback<Exclude<ImageContentProps["onClick"], undefined>>(
        event => {
            event.stopPropagation();
            if (onClickType === "action") {
                onClick?.();
            } else if (onClickType === "enlarge") {
                openLightbox();
            }
        },
        [onClick, onClickType, openLightbox]
    );

    const hasClickHandler = (onClickType === "action" && onClick) || onClickType === "enlarge";
    const sharedContentProps: ImageContentProps = {
        style,
        tabIndex: tabIndex ?? 0,
        onClick: hasClickHandler ? onImageClick : undefined,
        altText
    };

    const content =
        type === "image" ? (
            <ImageUi.ContentImage
                className={classNames({ "img-thumbnail": displayAs === "thumbnail" })}
                image={previewMode ? image : processImageLink(image, displayAs)}
                height={height}
                heightUnit={heightUnit}
                width={width}
                widthUnit={widthUnit}
                {...sharedContentProps}
            />
        ) : (
            <ImageUi.ContentIcon icon={image} isGlyph={type === "glyph"} size={iconSize} {...sharedContentProps} />
        );

    if (renderAsBackground) {
        return (
            <ImageUi.BackgroundImage
                className={className}
                style={style}
                image={image}
                height={height}
                heightUnit={heightUnit}
                width={width}
                widthUnit={widthUnit}
                onClick={event => {
                    event.stopPropagation();
                    onClick?.();
                }}
            >
                {backgroundImageContent}
            </ImageUi.BackgroundImage>
        );
    }

    return (
        <ImageUi.Wrapper
            className={className}
            responsive={responsive}
            type={type}
            hasImage={image !== undefined && image.length > 0}
        >
            {content}
            {!previewMode && lightboxIsOpen && (
                <Lightbox isOpen={lightboxIsOpen} onClose={onCloseLightbox}>
                    {type === "image" ? cloneElement(content, { image, onClick: undefined }) : content}
                </Lightbox>
            )}
        </ImageUi.Wrapper>
    );
};
