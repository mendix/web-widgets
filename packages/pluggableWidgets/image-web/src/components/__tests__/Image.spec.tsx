import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import { forwardRef } from "react";
import { ModalProps } from "react-overlays/esm/Modal";
import { Image, ImageProps } from "../Image/Image";

jest.mock("../../assets/ic24-close.svg", () => "close-button-icon-svg");

jest.mock("react-overlays/Modal", () =>
    forwardRef((props: ModalProps, ref) => {
        if (!props.show) return null;
        const MockName = "react-overlays-modal-mock";
        // The backdrop is rendered somewhere else in a portal, but for testing sake we put it here since we also mock.
        const BackdropMockName = "react-overlays-modal-backdrop-mock";
        return (
            // @ts-expect-error lower case custom name to make clear it's a mock
            <MockName {...props} ref={ref}>
                {props.children}
                {/* @ts-expect-error lower case custom name to make clear it's a mock */}
                <BackdropMockName>{props.renderBackdrop?.({ onClick: jest.fn(), ref: jest.fn() })}</BackdropMockName>
            </MockName>
        );
    })
);

const imageProps: ImageProps = {
    class: "",
    type: "image",
    image: "https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg",
    height: 300,
    heightUnit: "pixels",
    width: 300,
    widthUnit: "pixels",
    iconSize: 0,
    responsive: true,
    onClickType: "action",
    displayAs: "fullImage",
    renderAsBackground: false,
    backgroundImageContent: null
};

const glyphiconProps: ImageProps = {
    class: "",
    type: "glyph",
    image: "glyphicon-asterisk",
    iconSize: 20,
    height: 0,
    heightUnit: "pixels",
    width: 0,
    widthUnit: "pixels",
    responsive: true,
    onClickType: "action",
    displayAs: "fullImage",
    renderAsBackground: false,
    backgroundImageContent: null
};

const iconProps: ImageProps = {
    class: "",
    type: "icon",
    image: "mx-icon mx-icon-asterisk",
    iconSize: 20,
    height: 0,
    heightUnit: "pixels",
    width: 0,
    widthUnit: "pixels",
    responsive: true,
    onClickType: "action",
    displayAs: "fullImage",
    renderAsBackground: false,
    backgroundImageContent: null
};

describe("Image", () => {
    let user: UserEvent;

    beforeEach(() => {
        user = userEvent.setup();
    });

    it("renders the structure with an image", () => {
        const image = render(<Image {...imageProps} />);
        expect(image.asFragment()).toMatchSnapshot();
    });

    it("renders the structure with an image and percentage dimensions", () => {
        const image = render(
            <Image {...imageProps} height={100} width={100} heightUnit="auto" widthUnit="percentage" />
        );
        expect(image.asFragment()).toMatchSnapshot();
    });

    it("renders the structure with a glyph icon", () => {
        const image = render(<Image {...glyphiconProps} />);
        expect(image.asFragment()).toMatchSnapshot();
    });

    it("renders the structure with an icon", () => {
        const image = render(<Image {...iconProps} />);
        expect(image.asFragment()).toMatchSnapshot();
    });

    it("renders the structure as a background image", () => {
        const image = render(
            <Image {...imageProps} renderAsBackground backgroundImageContent={<div>Image content</div>} />
        );
        expect(image.asFragment()).toMatchSnapshot();
    });

    describe("when the onClickType is action", () => {
        it("calls the onClick when clicking on an image", async () => {
            const onClickMock = jest.fn();
            const { getByRole } = render(<Image {...imageProps} onClick={onClickMock} onClickType="action" />);
            const image = getByRole("button");
            expect(image).toBeInTheDocument();
            await user.click(image);
            expect(onClickMock).toHaveBeenCalled();
        });

        it("has tabindex if there is an action with OnClick", () => {
            const onClickMock = jest.fn();
            const { getByRole } = render(
                <Image {...imageProps} onClick={onClickMock} onClickType="action" tabIndex={1} />
            );
            const image = getByRole("button");
            expect(image.tabIndex).toBe(1);
        });

        it("has no tabindex if there is no action with OnClick", () => {
            const { getByRole } = render(<Image {...imageProps} />);
            const image = getByRole("img");
            expect(image.tabIndex).toBe(-1);
        });

        it("calls the onClick when clicking on a glyph icon", async () => {
            const onClickMock = jest.fn();
            const { getByRole } = render(<Image {...glyphiconProps} onClick={onClickMock} onClickType="action" />);
            const icon = getByRole("button");
            expect(icon).toBeInTheDocument();
            await user.click(icon);
            expect(onClickMock).toHaveBeenCalled();
        });

        it("calls the onClick when clicking on an icon", async () => {
            const onClickMock = jest.fn();
            const { getByRole } = render(<Image {...iconProps} onClick={onClickMock} onClickType="action" />);
            const icon = getByRole("button");
            expect(icon).toBeInTheDocument();
            await user.click(icon);
            expect(onClickMock).toHaveBeenCalled();
        });
    });

    describe("when the onClickType is enlarge", () => {
        it("shows a lightbox when the user clicks on the image", async () => {
            const { container, getByRole } = render(<Image {...imageProps} onClickType="enlarge" />);
            expect(container.querySelector(".mx-image-viewer-lightbox")).not.toBeInTheDocument();
            const image = getByRole("button");
            expect(image).toBeInTheDocument();
            await userEvent.click(image);
            const lightbox = container.querySelector(".mx-image-viewer-lightbox-backdrop");
            expect(lightbox).toBeInTheDocument();
        });

        it("closes the lightbox when the user clicks on the close button after opening it", async () => {
            const { container, getByRole } = render(<Image {...imageProps} onClickType="enlarge" />);
            const image = getByRole("button");
            expect(image).toBeInTheDocument();
            await userEvent.click(image);
            const lightbox = container.querySelector(".mx-image-viewer-lightbox-backdrop");
            expect(lightbox).toBeInTheDocument();
            const closeButton = container.querySelector(".close-button");
            expect(closeButton).toBeInTheDocument();
            await userEvent.click(closeButton!);
            expect(lightbox).not.toBeInTheDocument();
        });
    });

    it("does not trigger on clicks from containers if clicked on the image", async () => {
        const onClickOuterMock = jest.fn();
        const onClickImageMock = jest.fn();
        const { getByRole } = render(
            <div onClick={onClickOuterMock}>
                <Image {...imageProps} onClickType="action" onClick={onClickImageMock} />
            </div>
        );
        const image = getByRole("button");
        expect(image).toBeInTheDocument();
        await user.click(image);
        expect(onClickImageMock).toHaveBeenCalledTimes(1);
        expect(onClickOuterMock).not.toHaveBeenCalled();
    });

    describe("when there is an accessibility alt text", () => {
        it("is set properly on an image", () => {
            const { getByRole } = render(<Image {...imageProps} altText="this is an awesome image" />);
            const image = getByRole("img");
            expect(image).toHaveAttribute("alt", "this is an awesome image");
        });

        it("is set properly on a glyphicon", () => {
            const { getByRole } = render(<Image {...glyphiconProps} altText="this is an awesome glyphicon" />);
            const icon = getByRole("img");
            expect(icon).toHaveAttribute("aria-label", "this is an awesome glyphicon");
            expect(icon).toHaveAttribute("role", "img");
        });

        it("is set properly on an icon", () => {
            const { getByRole } = render(<Image {...iconProps} altText="this is an awesome icon" />);
            const icon = getByRole("img");
            expect(icon).toHaveAttribute("aria-label", "this is an awesome icon");
            expect(icon).toHaveAttribute("role", "img");
        });
    });

    describe("when there is no accessibility alt text", () => {
        it("nothing is set on an image", () => {
            const { getByRole } = render(<Image {...imageProps} />);
            const image = getByRole("img");
            expect(image).not.toHaveAttribute("alt");
        });

        it("nothing is set on a glyphicon", () => {
            const { getByRole } = render(<Image {...glyphiconProps} />);
            const icon = getByRole("img");
            expect(icon).not.toHaveAttribute("aria-label");
        });

        it("nothing is set on an icon", () => {
            const { getByRole } = render(<Image {...iconProps} />);
            const icon = getByRole("img");
            expect(icon).not.toHaveAttribute("aria-label");
        });
    });

    describe("when showing an image as a thumbnail", () => {
        it("includes the thumb=true URL param in the image", () => {
            const { getByRole } = render(<Image {...imageProps} displayAs="thumbnail" />);
            const image = getByRole("img") as HTMLImageElement;
            expect(image.src).toContain("thumb=true");
        });

        it("does not include the thumb=true URL param in the lightbox image", async () => {
            const { getByRole, getAllByRole } = render(
                <Image {...imageProps} displayAs="thumbnail" onClickType="enlarge" />
            );
            const image = getByRole("button") as HTMLImageElement;
            expect(image.src).toContain("thumb=true");
            await user.click(image);

            const allImages = getAllByRole("img") as HTMLImageElement[];
            expect(allImages.length).toBeGreaterThanOrEqual(2);
            expect(allImages[0].src).not.toContain("thumb=true");
            expect(allImages[1].src).not.toContain("thumb=true");
        });
    });

    describe("when showing as a background image", () => {
        it("shows the content", () => {
            const { getByText } = render(
                <Image {...imageProps} renderAsBackground backgroundImageContent={<div>Image content</div>} />
            );
            expect(getByText("Image content")).toBeInTheDocument();
        });

        it("properly handles on click event if configured by the user", async () => {
            const onClickMock = jest.fn();
            const { container } = render(
                <Image
                    {...imageProps}
                    renderAsBackground
                    backgroundImageContent={<div>Image content</div>}
                    onClick={onClickMock}
                    onClickType="action"
                />
            );
            const backgroundImage = container.querySelector(".mx-image-viewer.mx-image-background");
            expect(backgroundImage).toBeInTheDocument();
            await user.click(backgroundImage!);
            expect(onClickMock).toHaveBeenCalledTimes(1);
        });
    });
});
