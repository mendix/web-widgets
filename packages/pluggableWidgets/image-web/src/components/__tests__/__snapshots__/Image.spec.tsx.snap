// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders the structure as a background image 1`] = `
<DocumentFragment>
  <div
    class="mx-image-viewer mx-image-background"
    style="height: 300px; width: 300px; background-image: url(https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg);"
  >
    <div>
      Image content
    </div>
  </div>
</DocumentFragment>
`;

exports[`Image renders the structure with a glyph icon 1`] = `
<DocumentFragment>
  <div
    class="mx-image-viewer mx-image-viewer-responsive mx-image-viewer-icon"
  >
    <span
      class="glyphicon-asterisk glyphicon"
      role="img"
      style="font-size: 20px;"
    />
  </div>
</DocumentFragment>
`;

exports[`Image renders the structure with an icon 1`] = `
<DocumentFragment>
  <div
    class="mx-image-viewer mx-image-viewer-responsive mx-image-viewer-icon"
  >
    <span
      class="mx-icon mx-icon-asterisk"
      role="img"
      style="font-size: 20px;"
    />
  </div>
</DocumentFragment>
`;

exports[`Image renders the structure with an image 1`] = `
<DocumentFragment>
  <div
    class="mx-image-viewer mx-image-viewer-responsive"
  >
    <img
      class=""
      role="img"
      src="https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg"
      style="height: 300px; width: 300px;"
    />
  </div>
</DocumentFragment>
`;

exports[`Image renders the structure with an image and percentage dimensions 1`] = `
<DocumentFragment>
  <div
    class="mx-image-viewer mx-image-viewer-responsive"
  >
    <img
      class=""
      role="img"
      src="https://pbs.twimg.com/profile_images/1905729715/llamas_1_.jpg"
      style="width: 100%;"
    />
  </div>
</DocumentFragment>
`;
